src/request/request.rs

```diff
@@ -97,46 +113,43 @@ impl<State> Request<State> {
     /// }
     /// ```
     pub fn query<'a>(&self, name: &'a str) -> QueryParam<'_, 'a> {
-        let query = self.data.parts.uri.query().unwrap_or("");
-
+        let query = self.meta.parts.uri.query().unwrap_or("");
         QueryParam::new(name, query)
     }
 
     /// Returns a thread-safe reference-counting pointer to the application
     /// state that was passed as an argument to the [`via::app`](crate::app::app)
     /// function.
     pub fn state(&self) -> &Arc<State> {
-        &self.data.state
+        &self.state
     }
 
     /// Returns a reference to the uri associated with the request.
     pub fn uri(&self) -> &Uri {
-        &self.data.parts.uri
+        &self.meta.parts.uri
     }
 
     /// Returns the HTTP version associated with the request.
     pub fn version(&self) -> Version {
-        self.data.parts.version
+        self.meta.parts.version
     }
 
     /// Consumes the request and returns the body.
-    pub fn into_body(self) -> RequestBody {
-        self.data.body
+    pub fn into_body(self) -> AnyBody<Box<Incoming>> {
+        self.body
     }
 
     /// Unwraps `self` into the Request type from the `http` crate.
-    pub fn into_inner(self) -> http::Request<Incoming> {
-        let (parts, body, _) = self.into_parts();
-        let body = body.into_inner();
-
+    pub fn into_inner(self) -> http::Request<AnyBody<Box<Incoming>>> {
+        let (parts, body) = self.into_parts();
         http::Request::from_parts(parts, body)
     }
 
-    /// Consumes the request and returns a tuple containing the boxed component
-    /// parts of the request, the body, and an `Arc` to the shared application
-    /// state.
-    pub fn into_parts(self) -> (Parts, RequestBody, Arc<State>) {
-        (self.data.parts, self.data.body, self.data.state)
+    /// Consumes the request and returns a tuple containing the component
+    /// parts of the request and the request body.
+    pub fn into_parts(self) -> (Parts, AnyBody<Box<Incoming>>) {
+        let meta = *self.meta;
```

Comment:
This is safe because we take ownership of request and `self.meta` is immutable and does not expose any `Rc` or `Arc` pointers.